### ISSUE TITLE
How to install with composer

### ISSUE THREAD

**Issue by Crell** [shlinkio/shlink#975]
#### How Shlink is set-up

* Shlink Version: .5.1
* PHP Version: 7.4.12
* How do you serve Shlink: Self-hosted
* Database engine used: N/a

#### Summary

The `bin/install` command is failing with a missing class in the ocramius/package-versions package.

#### Current behavior / How to reproduce

```
git clone https://github.com/shlinkio/shlink.git
cd shlink
composer install --ignore-platform-reqs
bin/install
```

Produces:

```
PHP Fatal error:  Uncaught Error: Class 'Composer\InstalledVersions' not found in /home/crell/Platformsh/Scratch/shlink/vendor/ocramius/package-versions/src/PackageVersions/Versions.php:36
Stack trace:
#0 /home/crell/Platformsh/Scratch/shlink/vendor/laminas/laminas-cli/src/autoload.php(20): PackageVersions\Versions::getVersion()
#1 /home/crell/Platformsh/Scratch/shlink/vendor/composer/autoload_real.php(69): require('/home/crell/Pla...')
#2 /home/crell/Platformsh/Scratch/shlink/vendor/composer/autoload_real.php(59): composerRequire0bc557bf35e89f8bc359654e028d2e41()
#3 /home/crell/Platformsh/Scratch/shlink/vendor/autoload.php(7): ComposerAutoloaderInit0bc557bf35e89f8bc359654e028d2e41::getLoader()
#4 /home/crell/Platformsh/Scratch/shlink/vendor/shlinkio/shlink-installer/config/container.php(20): require_once('/home/crell/Pla...')
#5 /home/crell/Platformsh/Scratch/shlink/vendor/shlinkio/shlink-installer/bin/run.php(11): include('/home/crell/Pla...')
#6 /home/crell/Platformsh/Scratch/shlink/bin/install(11): require('/home/crell/Pla in /home/crell/Platformsh/Scratch/shlink/vendor/ocramius/package-versions/src/PackageVersions/Versions.php on line 36
```

I tried with both composer 1.10 and 2.0.8, same result.

#### Context

I am investigating Shlink to template-ify it for quick-install on Platform.sh, a PaaS host (where I work).  IN practice, that means Docker is not viable (we don't use Docker) and build a dist just to unpack it again immediately seems wasteful.  Most of our PHP apps install through straight composer, which is what I'm trying to do here.  I have to use the `--ignore-platform-reqs` because otherwise it fails to build since I don't have swoole installed, even though I don't plan to use the swoole version.

**Comment by acelaya** [comment]
Hey!

Up until now, the only officially supported method to install shlink is using the pregenerated dist files, like this one https://github.com/shlinkio/shlink/releases/download/v2.5.1/shlink_2.5.1_dist.zip

I'm open to explore other options if there's a use case for it, but I don't completely understand the sentence "build a dist just to unpack it again immediately" when the dist file is already generated 🤔. I'm probably missing something.

Other than that, those dist files are currently generated with composer v2.0.*

**Comment by Crell** [comment]
All our projects are fully Git-managed.  That's core to our design.  (GitOps and all that.)  We have starter kits for a wide variety of projects, but because they're Git-managed we do not commit vendor and other dependencies.  For example, here's the Symfony 5 template: https://github.com/platformsh-templates/symfony5, which then has `composer install` run on it each time it's deployed.  That makes it easy for users to self-update.  It's also easier for us to keep the starter kits up to date if we can just pull from an upstream Git repository to get the latest tagged version.

I could try unpacking the dist file and dumping that into the template, but that would make upgrading more of a chore for users.  (I still have to figure out how to make the installer work with a read-only file system for code, but that's step 2.)

Install-via-composer is a very common install method these days, so I'm surprised it's not supported.

**Comment by acelaya** [comment]
Well, sadly it's not as simple as `git clone && composer install`. Even without this error (which by the way, is caused by the use of `--ignore-platform-reqs`), there are a few other considerations to have in mind.

For example, after clonning the project it includes several dev assets that will only interfeer in prod envs (you can take a look at https://github.com/shlinkio/shlink/blob/develop/build.sh to see the steps needed to end up with a production-ready package).

Then, the project supports a few runtimes, which mean it has different dependencies that you don't always need, which in turn have their own dependencies (naming swoole, which is not required for your use case).

Having a controlled env where the dependencies are fetched and packaged together with the project, simplifies this and allows to offer a dist file that "just works". I have tried in the past to make some of these "optional", and it has proved to be trickier than it seems (there's still open tickets related with this topic https://github.com/shlinkio/shlink/issues/874).

And the final reason is that this approach has worked fine so far without too many people complaining (I can only think of one person which tried something similar in the past).

I agree that *install-via-composer* can be convenient, but that fits mostly 3 use cases, that I can think about:

* Install packages or frameworks: You named symfony, but shlink does not fall in the same category, as it's not a framework, but a service/app.
* Install projects using `create-project`: This was the first method used to install Shlink, many time ago, but again, it fets better when installing skeletons and such. Again, not exactly the same.
* Install private services: When the package supports a single set of dependencies and expects to run in one way only, because it's your company project. Shlink does not fit here either.

That said, I still think it should be possible to do what you are trying. and I think it's more or less reasonable, but it will require some manual work at first.

If you use composer 2, try using `--ignore-platform-req=swoole` instead of `--ignore-platform-reqs`, as this is what's making the wrong dependencies to be resolved, and let's continue from there.

**Comment by Crell** [comment]
Thanks, the more narrow ignore command worked!

I was able to get through the installer, although the file it generated conflates environment-specific bits (database creds, domain, etc.) with non-environment-specific bits.  Unfortunately that is not PaaS-friendly, so we would have to hack it and the installer quite extensively to get it to run in a multi-environment PaaS.  That's before we even talk about setting up the UI for it.

I'm afraid this isn't going to work, so we'll have to skip supporting this app. :disappointed:  Thanks for your help.

**Comment by acelaya** [comment]
Making it pick config from env vars should be easy. It is already done for the docker image by dynamically including [this config file](https://github.com/shlinkio/shlink/blob/develop/docker/config/shlink_in_docker.local.php) inside the `config/autoload` folder.

With that you wouldn't need to use the installer at all (it doesn't fit your use case anyway), but just pass env vars as defined here https://shlink.io/documentation/install-docker-image/#supported-env-vars

At the moment this file is only picked up when building the docker image ( see https://github.com/shlinkio/shlink/blob/develop/Dockerfile#L74), but I've been considering supporting env vars by default regardless the approach, so I'm open to do that.

**Comment by Crell** [comment]
Ah ha!  Yes, that would be very helpful.

Ideally, users would be able to set the non-environment-specific bits in the database (the file system is read only) and the environment-specific bits (database, URL, etc.) are just read from the environment directly and the user is never bothered with them.

I haven't tried out the UI component yet, but all of the same caveats apply there as well.  In the very ideal, the UI could also trigger the installer so that users didn't need to go to the CLI to do so.  I don't know how feasible that is, though.

**Comment by acelaya** [comment]
Shlink could be considered mostly stateless, but there are a few things to take into consideration.

* The logs: When running with swoole, the project just sends logs to the standard output, but when running with FastCGI (behind nginx/apache) it generates logs in the filesystem. This shouldn't be hard to customize though, as it uses monolog, so it should be possible to send logs somewhere else.
    How do you handle logs in Platform.sh for other services?
* The proxies: The project generates some proxies, including doctrine/orm ones. This behavior could be disabled, but it would have a small impact in performance, as they would need to be regenerated in memory in every request.
* The locks: The project generates some filesystem-based locks to make sure some tasks are not run in parallel more than once. However it also supports locks over an external redis, in case shlink is clusterized.
    The project uses symfony/lock for this, which has support for a few other locking backends, if redis is also not an option.

So basically you would need to request all the config to the user somehowe, and then funnel it to Shlink through env vars. Everything is treated the same, from database connection params, to customizations like what's the user's domain or what status code (301/302) should be used for redirects.

This would allow you skip the `bin/install` step, which is more suitable for people deploying on a VPS or similar. The Docker image does not use it, for example.

However, there's still some tasks that will require access to the CLI, mostly related with API key handling. API keys can only be generated through the CLI for security reasons.

**Comment by Crell** [comment]
1. We don't have a custom logging facility at the moment.  Just logging with `trigger_error()` et al is fine.  You can also log yourself to a directory using Monolog, as long as that directory is flagged as a writeable mount, meaning it cannot contain any files that come from Git.  (Symfony and Laravel do that, no issue.)
2. If those can be generated without database access, they can be done in the build step and written to a directory that is read-only once it's deployed.  That's preferable.  If not, it can be done at runtime as long as, again, the directory where they're written to is exclusively runtime-write and has no files that come from Git.
3. Redis is fine, and easy to setup on our end as long as the credentials for it can be fed in via env vars.  Locking to disk is also acceptable as long as, again, that directory is exclusively a runtime-write directory with no mixing of Git-provided files.

Feeding in DB and Redis creds via env vars is fine, and it looks like that would work after all.  The tricky part is the combined config file.  In the ideal case, that would be two separate files in separate directories, one of which gets ignored if the installer detects the env vars it needs.  Then the user config can still be written out based on the installer to a writeable config directory that is separate from a read-only config directory, if any.  In fact, at that point you may want to use a `.env` file for the info you expect to come from the environment, as that fallback logic is fairly well-understood and there's a zillion existing libraries to handle it for you.  We don't currently have a way to force the user to set app-specific env vars in advance.

Users do have CLI access, we just try to not make it a requirement for installing the application.  What we've done for some templates is have a pre-canned CLI command run during the first deploy only to run an install process, which sometimes includes printing out an API key into the deploy log, something we then document.  That would *probably* work here as well.  It's the same we do for eZ Platform^W^W Ibexa and Nextcloud, for instance.

I've written more about installers in the past, here: https://platform.sh/blog/2020/installers-that-dont-suck/

(The ideal is that we could get a template going that is self-serve enough that we can include it here: https://platform.sh/marketplace/ .  And then that button can be placed on other sites, too, like the Shlink website, to make it as easy as possible for users to spin up a Shlink instance read-to-go.)

**Comment by acelaya** [comment]
1, 2 and 3 are in all cases generated in folders which do not have anything else. They are part of the git repo though, including a `.gitignore` file. Could that be a problem? (This is an example https://github.com/shlinkio/shlink/tree/develop/data/log)

There's actually a fourth folder that I forgot to mention, where some assets are cached (mostly configs once parsed), but again, it is a folder which isn't used for anything else.

> The tricky part is the combined config file. In the ideal case, that would be two separate files in separate directories

There's no problem. The config can be split in as many files as needed. In the case of docker I use only one just for conveniency, but the project uses [this package](https://github.com/laminas/laminas-config-aggregator#laminas-config-aggregator) to merge config files at runtime from many sources.

The only thing is that for production it dumps a single file with the merged result, so that in next requests it just includes it and that's it.

That can have two considerations for your use case:

* It needs to be able to write in the `data/cache` folder, but it's the same as 1, 2 and 3, so I imagine it should be ok (this is the one I forgot to mention).
* If the user changes the config later on this one file, the cached config has to be deleted for the changes to take effect. This could be solved by disabling this config caching mechanism, but it would have a performance impact as it would need to "require" all the config files and merge them on every request.

> In fact, at that point you may want to use a `.env` file

There was a time in which Shlink used an env file in dev mode, but I replaced it with local config files.

Having some optional way to support this again shouldn't be a big issue. However I also see the problem that this kind of mechanism (env vars defined in a `.env` file) are not usually meant for production.

> What we've done for some templates is have a pre-canned CLI command run during the first deploy only to run an install process, which sometimes includes printing out an API key into the deploy log, something we then document. That would probably work here as well.

Yep, this would work. Running `bin/cli api-key:generate` would generate an admin API key and print it in the standard output.

That all said, it would probably help me understand what's missing if I could experiment a bit with the platform and try to deploy Shlink there.

Putting myself in the position of a platform.sh user will help me see what should be provided and how. Is there any option I could get an account to play with?

P.S.: Sorry for the delay on my response. I had some things to tackle this week.

**Comment by Crell** [comment]
> P.S.: Sorry for the delay on my response. I had some things to tackle this week.

No worries.  Volunteer projects don't have schedules.  Been there, done that, have the T-shirt. :smile: 

A `.gitignore` file is no problem.  It's actually recommended to avoid people accidentally committing files that would then get hidden in production.

In practice, requiring multiple config files has almost no measurable overhead these days.  PHP is quite fast, most storage is SSD, and with the amount of OS caching involved including 2-3 files instead of one is almost unmeasurable.

You're correct that `.env` files are not meant to be used in production.  However, if the app is configurable via real environment variables then there's no issue.  What we do for a number of applications is something like this tiny bash script that runs before the app starts and maps our env vars to the ones the application expects:  https://github.com/platformsh-templates/akeneo/blob/master/.environment#L14

As long as the app can read environment-specific configuration from real env vars, that works great and then .env works for local.

> Putting myself in the position of a platform.sh user will help me see what should be provided and how. Is there any option I could get an account to play with?

Certainly.  Email me at larry at my company and I'll setup a dev account and give you access to it.  We ideally prefer if vendors can maintain their own template and we just reference it, so all users can spin up a project with it with a few clicks.  Platform.sh is a bit unusual if you're used to traditional hosting but it's quite powerful.  If you have any questions or need help understanding something, the best route is to stop by https://chat.platform.sh/ and ask in our public channel.  I'm generally around US business hours (same name there), but the rest of the team is good at answering, too.

Thanks for being so open to collaborating!

**Comment by acelaya** [comment]
> We ideally prefer if vendors can maintain their own template and we just reference it.

I'm afraid I'm not so sure about this.

I can help figuring out how to make it work, and spend some time testing and adapting the project so that it is possible to do it, but after that, I can't commit to maintain it.

It's not something that's going to benefit the majority of Shlink's user base, and it's a payed service.
